### PR TITLE
Fortunac/general loop invariant checker

### PR DESCRIPTION
**Pull request by fortunac**
- Simplifies the loop invariant checker to handle more general cases
  - Rather than finding the loop guard and creating constraints from those, we find the weakest precondition for the loop itself.
  - `invariant /\ (invariant => ite(!loop_guard, precondition of the exit, wp(loop_body, invariant))`
- Adds more test cases:
  - A loop with a nondeterministic number of iterations
  - A do while loop
  - A loop with a break
  - A non-terminating loop

### CODE REVIEW

File: wp/lib/bap_wp/src/precondition.ml [hunk 1]
```diff
@@ -901,8 +901,12 @@ let visit_graph (env : Env.t) (post : Constr.t)
         let dst = G.Edge.dst e in
         debug "Entering back edge from\n%sto\n%s\n%!"
           (G.Node.to_string src) (G.Node.to_string dst);
-        let handler = Env.get_loop_handler env in
-        post, handler env post ~start:dst g
+        let tid = dst |> G.Node.label |> Term.tid in
```
Comment (codyroux) [inline wp/lib/bap_wp/src/precondition.ml:904]: I'm wondering, is the general approach of how we handle this documented anywhere?

I guess this is more of a new issue than anything else.

File: wp/lib/bap_wp/src/precondition.ml [hunk 2]
```diff
@@ -1397,47 +1401,36 @@ let in_loop (loop : Graphs.Ir.Node.t group) (tid : Tid.t) : bool =
    - for each node in the SCC, find its parent in the dominator tree
    - if the parent node is not in the original SCC, it is an exit node.
 
-   The exit node of a loop is the first node outside of the SCC in the control
-   flow graph. This should be the node the loop header points to that isn't a
-   part of the loop.
+   The exit node of a loop is a node outside of the SCC in the control flow
+   graph. This node will always be visited when exiting a loop.
```
Comment (codyroux) [inline wp/lib/bap_wp/src/precondition.ml:1405]: Not sure what that last sentence means. Is it "will have already been visited"?

File: wp/lib/bap_wp/src/precondition.ml (hunk 1 above)
Comment (fortunac) [reply wp/lib/bap_wp/src/precondition.ml:904]: I could either put this in the comments for the loop invariant checker or in the design decisions markdown file. Leaning towards the design decisions.

File: wp/lib/bap_wp/src/precondition.ml (hunk 2 above)
Comment (fortunac) [reply wp/lib/bap_wp/src/precondition.ml:1405]: The execution of the program will always go through the exit node if we are on the path that contains the loop.

Comment (codyroux) [reply wp/lib/bap_wp/src/precondition.ml:1405]: Is this true though? What if a loop has 2 exits?

Comment (fortunac) [reply wp/lib/bap_wp/src/precondition.ml:1405]: Hmm.. I guess not always true. I only had a few samples with multiple exits, and this always returned the node that all the exits pointed to. I can't say for other loop shapes.

Comment (codyroux) [reply wp/lib/bap_wp/src/precondition.ml:1405]: So we don't actually know what this code does?

File: wp/lib/bap_wp/src/precondition.ml [hunk 3]
```diff
@@ -1397,47 +1401,36 @@ let in_loop (loop : Graphs.Ir.Node.t group) (tid : Tid.t) : bool =
...
+  Seq.find_map (Group.enum loop) ~f:(fun n ->
       let* parent = Tree.parent dom_tree n in
-      if Group.mem group parent then
+      if Group.mem loop parent then
         None
       else
         Some parent)
 
 (* Gets the precondition of the exit node of a loop. *)
 let exit_pre (env : Env.t) (post : Constr.t) (node : Graphs.Ir.Node.t)
```
Comment (codyroux) [inline wp/lib/bap_wp/src/precondition.ml:1420]: 👍

Comment (codyroux) [reply wp/lib/bap_wp/src/precondition.ml:1420]: Though I guess we could do this *as we hit the nodes* during the loop analysis, but I'm not sure that would improve the code.

File: wp/lib/bap_wp/src/precondition.ml [hunk 4]
```diff
@@ -1468,127 +1466,27 @@ let loop_unroll (num_unroll : int) : Env.loop_handler =
...
-  Graphlib.depth_first_search (module Graphs.Ir)
-    ~enter_node ~enter_edge ~init:graph graph
-
-(* Gets a set of vars that get updated in the body of a loop. *)
-let loop_vars (sub : Sub.t) : Var.Set.t =
-  let visitor =
-    (object inherit [Var.Set.t] Term.visitor
-      method! visit_def def vars =
-        Var.Set.add vars (Def.lhs def)
-    end)
```
Comment (codyroux) [inline wp/lib/bap_wp/src/precondition.ml:1580]: Oof that feels nice

**Comment by codyroux** [comment]
This is neato burrito! Very nice stuff.

File: wp/lib/bap_wp/src/precondition.ml (hunk 3 above)
Comment (fortunac) [reply wp/lib/bap_wp/src/precondition.ml:1420]: Are you suggesting a hook in `visit_graph` where we calculate the parent as we enter the node, and then if it is a parent in the post-dominator tree, we recursively call `visit_graph` to update the env?

Comment (codyroux) [reply wp/lib/bap_wp/src/precondition.ml:1420]: Yes, that's what I'm suggesting, as maybe a `jmp_spec` or whatever we call it.

Again: not sure that would make the code cleaner in any way, so I wouldn't sweat it.

**Pull request merged**